core/src/main/java/org/jruby/RubyClass.java

```diff
@@ -1496,6 +1492,18 @@ public synchronized void reify(String classDumpDir, boolean useChildLoader) {
         }
     }
 
+    private String getReifiedJavaClassName() {
+        final String basePackagePrefix = "rubyobj.";
+        if (getBaseName() == null) { // anonymous Class instance: rubyobj.Class$0x1234abcd
+            return basePackagePrefix + anonymousMetaNameWithIdentifier().replace(':', '$');
+        }
+        final String identityHash = Integer.toHexString(System.identityHashCode(this));
+        final CharSequence name = StringSupport.replaceAll(getName(), "::", ".");
+        // we need to include a Class identifier in the Java class name, since a Ruby class might be dropped
+        // (using remove_const) and re-created in which case using the same name would cause a conflict...
+        return basePackagePrefix + identityHash + '.' + name; // TheFoo::Bar -> rubyobj.320efdf5.TheFoo.Bar
```

Comment:
and while this works "better" (well except for 'breaking' the `rubyobj.` naming convention) it might still be better to just add a (counter) suffix. given patching of a reified class should re-generate the Java class which means there needs to be more classloader-fu or we simply add a `v2` suffix and keep the existing loader (previous class is no longer used so should GC away, eventually)